### PR TITLE
add ModifyDN() - moddn/modrdn

### LINKED ISSUE #63: Missing ModDN / ModRDN functionality
see https://github.com/go-ldap/ldap/pull/54

### PR DESCRIPTION
**Pull request by vetinari**
fixes #63
This implements the ModifyDNRequest from RFC 4511

### CODE REVIEW

**Comment by vetinari** [comment]
@liggitt what about merging this one and https://github.com/go-ldap/ldap/pull/62? This PR is already running in our local version in production

**Comment by iamthemuffinman** [comment]
+1

File: moddn.go [hunk 1]
```diff
@@ -0,0 +1,95 @@
...
+		}
+	} else {
+		log.Printf("Unexpected Response: %d", packet.Children[1].Tag)
+	}
+
+	l.Debug.Printf("%d: returning", messageID)
+	return nil
+}
+
+// vim: ts=4 sw=4 noexpandtab nolist
```
Comment (liggitt) [inline moddn.go]: drop this line

File: moddn.go [hunk 2]
```diff
@@ -0,0 +1,95 @@
...
+)
+
+type ModifyDNRequest struct {
+	DN           string
+	NewRDN       string
+	DeleteOldRDN bool
+	NewSuperior  string
+}
+
+// set "newSup" to empty string for just renaming, to move w/o renaming the "rdn" must be the first
```
Comment (liggitt) [inline moddn.go]: lead the comment with function name, expand abbreviations, split parameter usage tips into separate lines/sentences (maybe even with examples?)

File: moddn.go [hunk 3]
```diff
@@ -0,0 +1,95 @@
...
+	if err != nil {
+		return err
+	}
+	if channel == nil {
+		return NewError(ErrorNetwork, errors.New("ldap: could not send message"))
+	}
+	defer l.finishMessage(messageID)
+
+	l.Debug.Printf("%d: waiting for response", messageID)
+	packet = <-channel
```
Comment (liggitt) [inline moddn.go]: needs to be updated for changes in https://github.com/go-ldap/ldap/pull/57:
```
	packetResponse, ok := <-channel
	if !ok {
		return NewError(ErrorNetwork, errors.New("ldap: channel closed"))
	}
	packet, err = packetResponse.ReadPacket()
	l.Debug.Printf("%d: got response %p", messageID, packet)
	if err != nil {
		return err
	}
```

**Comment by liggitt** [comment]
just a few comments, then LGTM, thanks

**Comment by jamesboswell** [comment]
+1  nice job @vetinari !

File: moddn.go [hunk 4]
```diff
@@ -0,0 +1,104 @@
...
+	DeleteOldRDN bool
+	NewSuperior  string
+}
+
+// NewModifyDNRequest creates a new request which can be passed to ModifyDN().
+//
+// To move an object in the tree, set the "newSup" to the new parent entry DN. Use an
+// empty string for just changing the object's RDN.
+//
+// For moving the object without renaming the "rdn" must be the first
```
Comment (jamesboswell) [inline moddn.go]: Example for moving without renaming?  Also a comma  "without renaming, the"

File: moddn.go [hunk 5]
```diff
@@ -0,0 +1,104 @@
...
+// RDN of the given DN.
+//
+// A call like
+//   mdnReq := NewModifyDNRequest("uid=someone,dc=example,dc=org", "uid=newname", true, "")
+// will setup the request to just rename uid=someone,dc=example,dc=org to
+// uid=newname,dc=example,dc=org.
+func NewModifyDNRequest(dn string, rdn string, delOld bool, newSup string) *ModifyDNRequest {
```
Comment (jamesboswell) [inline moddn.go]: For future documentation clarity (godoc) it wouldn't hurt to have 3 short examples:

1. Rename without move
2. Rename with move
3. Move only

Or just actual `Example` functions in a moddn_test.go file?  (https://blog.golang.org/examples)

Comment (vetinari) [inline moddn.go]: made an moddn_test.go with the 3 examples

**Comment by iavael** [comment]
Hi. Is any progress expected in this PR? Looking forward for this feature.

**Comment by oldCoderException** [comment]
Awesome to see this.  Good work!   When will it be released?  I'm struggling with RDN renaming (login change) right now.

**Comment by oldCoderException** [comment]
Perhaps somehow I'm viewing an older version or something, but on line 64 of moddn.go, shouldn't l.nextMessageID be the function call l.nextMessageID() ??
As it is I got a panic: Invalid type func() int64, expected {u|}int{64|32|16|8}
from gopkg.in/asn1-ber.v1/ber.go:491 +0x244
Changing to the function call resulted in success.
Again, thanks for this.  Very needed function in LDAP!  Excellent work.

**Comment by vetinari** [comment]
@oldCoderException right, the () were missing. probably happened while porting my production tree to this one

**Comment by oldCoderException** [comment]
Don't mean to be a pain but any idea when this makes it into the baseline?  We've been using it in production for a few months now.

Thanks!

**Comment by liggitt** [comment]
No concerns, though I haven't tried the function against a real server

**Comment by jamesboswell** [comment]
I recently tested this against production LDAP using my [fork](https://github.com/jamesboswell/ldap) which cherry-picks this PR.

I needed to migrate many hundreds of users uid to a new login schema (pass-through to AD).  It is a component of the DN in this environment.  @vetinari moddn PR worked perfectly.

@vetinari FYI I had to drop your README modifications to this PR into my fork, you need to update it to make it merge ready

@johnweldon  @liggitt  any chance this can be merged at some point before it's 2 year anniversary?

**Comment by vb4life** [comment]
I would ***LOVE*** to have this feature merged, since I'm developing a web-based user management system using go, and renaming a user is quite common.

**Comment by liggitt** [comment]
rebased as #149

**Comment by liggitt** [comment]
merged in #149

**Pull request closed without merging**